{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate pymol script\n",
    "The scripts can be found in `Data/SARS-CoV-2_spike/Structures/` as `script_fixed.pml` and `script_parallel.pml`\n",
    "\n",
    "# Important\n",
    "The command line tool `muscle` is used for multiple sequence alignment. Make sure the executable can be found in the `PATH`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import json\n",
    "from io import StringIO\n",
    "from urllib import request\n",
    "from subprocess import Popen, PIPE\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.PDB import PDBList\n",
    "\n",
    "DATA_DIR = \"Data\"\n",
    "REFERENCE = \"EPI_ISL_402124\"\n",
    "\n",
    "PDB_SERVER = PDBList(verbose=False)\n",
    "PDB_ID = { \n",
    "    \"SARS-CoV-2\": {\n",
    "        \"spike\": \"7DF3\"\n",
    "    } \n",
    "}\n",
    "PROTEIN_DOMAIN = {\n",
    "    \"SARS-CoV-2\": {\n",
    "        \"spike\": {\n",
    "            \"RBD\": [s for s in range(319, 541 + 1)],\n",
    "            \"NTD\": [s for s in range(13, 303 + 1)],\n",
    "            \"S2\": [s for s in range(686, 1273 + 1)],\n",
    "            \"FP\": [s for s in range(816, 855 + 1)],\n",
    "            \"HR1\": [s for s in range(920, 970 + 1)],\n",
    "            \"HR2\": [s for s in range(1163, 1202 + 1)]\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "SITES_COLOR = {\n",
    "    \"fixation\": \"gold\",\n",
    "    \"parallel\": \"red\", \n",
    "    \"paraFix\": \"red\",\n",
    "}\n",
    "DOMAIN_COLOR = {\n",
    "    \"Seq\": \"0xBCBDBD\",\n",
    "    \"NTD\": \"0x9BCAC8\",\n",
    "    \"RBD\": \"0xC6716B\",\n",
    "    \"S2\": \"0x9EC9A1\",\n",
    "    \"FP\": \"0xBAD3E1\",\n",
    "    \"HR1\": \"0xF0D0CE\",\n",
    "    \"HR2\": \"0xF0D0CE\"\n",
    "    \n",
    "#     \"NTD\": \"blue\",\n",
    "#     \"RBD\": \"magenta\",\n",
    "#     \"S2\": \"green\",\n",
    "#     \"FP\": \"orange\",\n",
    "#     \"HR1\": \"purple\",\n",
    "#     \"HR2\": \"purple\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SARS-CoV-2 spike protein\n",
    "\n",
    "Generate pymol script for SARS-CoV-2 spike protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "virus = \"SARS-CoV-2\"\n",
    "protein = \"spike\"\n",
    "\n",
    "inputDir = os.path.join(DATA_DIR, virus + \"_\" + protein)\n",
    "\n",
    "outdirPath = os.path.join(inputDir, \"Structures\")\n",
    "if os.path.exists(outdirPath):\n",
    "    shutil.rmtree(outdirPath)\n",
    "os.mkdir(outdirPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the corresponding PDB file\n",
    "pdbFilePath = PDB_SERVER.retrieve_pdb_file(\n",
    "    pdb_code=PDB_ID[virus][protein],\n",
    "    file_format=\"pdb\",\n",
    "    pdir=outdirPath\n",
    ")\n",
    "\n",
    "# Find the 'DBREF' entry in the PDB file for sites mapping\n",
    "sitesMapping = []\n",
    "with open(pdbFilePath, 'r') as f:\n",
    "    monomer = True\n",
    "    monomer_ids = set()\n",
    "    previous_e = -1\n",
    "    completeRef = False\n",
    "    for line in f:\n",
    "        if line.startswith(\"DBREF\"):\n",
    "            if line.startswith(\"DBREF1 \"):\n",
    "                completeRef = False\n",
    "                chain_id, pdb_s, pdb_e = (\n",
    "                    line[12],\n",
    "                    int(line[14:18]),\n",
    "                    int(line[20:24]) + 1,\n",
    "                )\n",
    "            elif line.startswith(\"DBREF2 \"):\n",
    "                completeRef = True\n",
    "                chain_id, ref_s, ref_e, uniprot_id = (\n",
    "                    line[12],\n",
    "                    int(line[45:55]),\n",
    "                    int(line[57:67]) + 1,\n",
    "                    line[18:40].strip()\n",
    "                )\n",
    "            if line.startswith(\"DBREF \"):\n",
    "                completeRef = True\n",
    "                chain_id, pdb_s, pdb_e, ref_s, ref_e, uniprot_id = (\n",
    "                    line[12],\n",
    "                    int(line[14:18]),\n",
    "                    int(line[20:24]) + 1,\n",
    "                    int(line[55:60]),\n",
    "                    int(line[62:67]) + 1,\n",
    "                    line[33:41].strip(),\n",
    "                )\n",
    "            if completeRef:\n",
    "                if previous_e > ref_s:\n",
    "                    monomer = False\n",
    "                previous_e = ref_e\n",
    "                if monomer:\n",
    "                    monomer_ids.add(chain_id)\n",
    "                    for pdb_p, ref_p in zip(range(pdb_s, pdb_e), range(ref_s, ref_e)):\n",
    "                        sitesMapping.append({\"uniprotSite\": ref_p, \"chain\": chain_id, \"pdbSite\": pdb_p})\n",
    "    sitesMappingAll = pd.DataFrame.from_records(sitesMapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download the uniprot sequence\n",
    "res = request.urlopen(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "uniprotSeq = SeqIO.read(StringIO(res.read().decode(\"utf-8\")), \"fasta\")\n",
    "uniprotSeq.id = uniprotSeq.description.split('|')[1]\n",
    "uniprotSeq.description = \"Reference\"\n",
    "\n",
    "# Combine the sequences for multiple sequence alignment\n",
    "sequences = [record for record in SeqIO.parse(os.path.join(inputDir, \"selected_AA.fasta\"), \"fasta\")]\n",
    "refAddedFilePath = os.path.join(outdirPath, \"refAdded.fasta\")\n",
    "SeqIO.write(\n",
    "    sequences=[*sequences, uniprotSeq],\n",
    "    handle=refAddedFilePath,\n",
    "    format=\"fasta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Multiple sequence alignment using muscle\n",
    "process = Popen(\n",
    "    [\"muscle\", \"-in\", refAddedFilePath, \"-diags\"],\n",
    "    stdout=PIPE,\n",
    "    stderr=PIPE\n",
    ")\n",
    "stdout, stderr = process.communicate()\n",
    "if process.returncode:\n",
    "    raise BaseException(\"muscle: \" + stderr.decode(\"UTF-8\"))\n",
    "    \n",
    "# Map the site numbering between PDB DBREF (uniprot) and the reference\n",
    "seqs = SeqIO.to_dict(SeqIO.parse(StringIO(stdout.decode(\"UTF-8\")), \"fasta\"))\n",
    "refSite, uniprotSite = 0, 0\n",
    "sitesMapping = []\n",
    "for n, (ref_aa, uniprot_aa) in enumerate(zip(seqs[REFERENCE].seq, seqs[uniprot_id].seq), start=1):\n",
    "    validSite = False\n",
    "    if ref_aa != \"-\":\n",
    "        refSite += 1\n",
    "        validSite = True\n",
    "    if uniprot_aa != \"-\":\n",
    "        uniprotSite += 1\n",
    "        validSite = True\n",
    "    if validSite:\n",
    "        sitesMapping.append({\"refSite\": refSite, \"uniprotSite\": uniprotSite, \"alignedSite\": n})\n",
    "sitesMapping = pd.DataFrame.from_records(sitesMapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sitesMappingAll = pd.merge(sitesMappingAll, sitesMapping, on=\"uniprotSite\", how=\"left\")\n",
    "# sitesMappingAll = sitesMappingAll.merge( \n",
    "#     genomeSitesMapping[genomeSitesMapping[\"product\"] == protein],\n",
    "#     left_on=\"refSite\",\n",
    "#     right_on=\"aaPos\",\n",
    "#     how=\"left\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprotSite</th>\n",
       "      <th>chain</th>\n",
       "      <th>pdbSite</th>\n",
       "      <th>refSite</th>\n",
       "      <th>alignedSite</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>1204</td>\n",
       "      <td>A</td>\n",
       "      <td>1204</td>\n",
       "      <td>1204</td>\n",
       "      <td>1204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>1205</td>\n",
       "      <td>A</td>\n",
       "      <td>1205</td>\n",
       "      <td>1205</td>\n",
       "      <td>1205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>1206</td>\n",
       "      <td>A</td>\n",
       "      <td>1206</td>\n",
       "      <td>1206</td>\n",
       "      <td>1206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>1207</td>\n",
       "      <td>A</td>\n",
       "      <td>1207</td>\n",
       "      <td>1207</td>\n",
       "      <td>1207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>1208</td>\n",
       "      <td>A</td>\n",
       "      <td>1208</td>\n",
       "      <td>1208</td>\n",
       "      <td>1208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1208 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprotSite chain  pdbSite  refSite  alignedSite\n",
       "0               1     A        1        1            1\n",
       "1               2     A        2        2            2\n",
       "2               3     A        3        3            3\n",
       "3               4     A        4        4            4\n",
       "4               5     A        5        5            5\n",
       "...           ...   ...      ...      ...          ...\n",
       "1203         1204     A     1204     1204         1204\n",
       "1204         1205     A     1205     1205         1205\n",
       "1205         1206     A     1206     1206         1206\n",
       "1206         1207     A     1207     1207         1207\n",
       "1207         1208     A     1208     1208         1208\n",
       "\n",
       "[1208 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sitesMappingAll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The sites (numbering based on EPI_ISL_402125) to be marked on the structure\n",
    "targetSites = {}\n",
    "chains = set()\n",
    "with open(os.path.join(inputDir, \"targetSites.json\")) as f:\n",
    "    targetSites = json.load(f)\n",
    "    \n",
    "for siteType, sites in targetSites.items():\n",
    "    sites = sitesMappingAll[sitesMappingAll[\"refSite\"].isin(sites)]\n",
    "    sites = sites[[\"chain\", \"pdbSite\"]].drop_duplicates()\n",
    "    targetSites[siteType] = sites\n",
    "    chains = chains.union(sites[\"chain\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "pmlSetView = \"set_view (\\\n",
    "     0.787351847,   -0.214461818,    0.577996790,\\\n",
    "     0.602337480,    0.067778848,   -0.795357168,\\\n",
    "     0.131397858,    0.974372327,    0.182546288,\\\n",
    "     0.000000000,    0.000000000, -539.911254883,\\\n",
    "   163.201431274,  163.200393677,  166.706909180,\\\n",
    "  -16535.880859375, 17615.703125000,  -20.000000000 )\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pmlScript = [\n",
    "    \"delete *\",\n",
    "    \"fetch {}\".format(PDB_ID[virus][protein]),\n",
    "    \"hide\",\n",
    "    \"bg_color white\",\n",
    "    \"show surface\",\n",
    "    \"set transparency, 0.9\",\n",
    "    \"set surface_color, grey\",\n",
    "    \"; \".join([\"show cartoon, c. \" + c for c in chains]),\n",
    "    \"color grey\"\n",
    "]\n",
    "\n",
    "for domain, sites in PROTEIN_DOMAIN[virus][protein].items():\n",
    "    sites = sitesMappingAll.loc[sitesMappingAll[\"refSite\"].isin(sites), [\"chain\", \"pdbSite\"]]\n",
    "    sel = \" or \".join([f\"c. {c} and i. {i}\" for _, c, i in sites.itertuples()])\n",
    "    col = DOMAIN_COLOR[domain]\n",
    "    pmlScript.extend([\n",
    "        f\"select {domain}, {sel}\",\n",
    "        f\"color {col}, {domain}\",\n",
    "        f\"set surface_color, {col}, {domain}\",\n",
    "        f\"set transparency, 0.75, {domain}\"\n",
    "#         f\"create Iso_{domain}, {domain}\",\n",
    "#         f\"show surface, Iso_{domain}\",\n",
    "#         f\"set surface_color, {col}, Iso_{domain}\",\n",
    "#         f\"set transparency, 0.1, Iso_{domain}\"\n",
    "    ])\n",
    "\n",
    "# Script for fixation sites\n",
    "for category, sites in targetSites.items():\n",
    "    if category == \"fixation\":\n",
    "        sel = \" or \".join([f\"c. {c} and i. {i}\" for _, c, i in sites.itertuples()])\n",
    "        col = SITES_COLOR[category]\n",
    "        pmlScript.extend([\n",
    "            f\"select {category}, {sel}\",\n",
    "            f\"color {col}, {category}\",\n",
    "            f\"show sphere, {category}\",\n",
    "            f\"disable {category}\"\n",
    "        ])\n",
    "        pmlScript.append(\"reset\")\n",
    "        pmlScript.append(pmlSetView)\n",
    "        # For the site's label\n",
    "        label = \" or \".join([f\"c. {c} and i. {i} and n. CA\" for _, c, i in sites.itertuples()])\n",
    "        pmlScript.extend([\n",
    "            \"set label_position,(5,7,100)\",\n",
    "#             \"set label_size, 10\",\n",
    "            \"set label_connector, true\",\n",
    "            \"set label_bg_color, white\",\n",
    "            \"set label_bg_transparency, 0.2\",\n",
    "            f\"label {label}, resi\\n\"\n",
    "        ])\n",
    "\n",
    "with open(os.path.join(outdirPath, \"script_fixed.pml\"), 'w') as f:\n",
    "    f.write(\"; \".join(pmlScript))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pmlScript = [\n",
    "    \"delete *\",\n",
    "    \"fetch {}\".format(PDB_ID[virus][protein]),\n",
    "    \"hide\",\n",
    "    \"bg_color white\",\n",
    "    \"show surface\",\n",
    "    \"set transparency, 0.9\",\n",
    "    \"set surface_color, grey\",\n",
    "    \"; \".join([\"show cartoon, c. \" + c for c in chains]),\n",
    "    \"color grey\"\n",
    "]\n",
    "\n",
    "for domain, sites in PROTEIN_DOMAIN[virus][protein].items():\n",
    "    sites = sitesMappingAll.loc[sitesMappingAll[\"refSite\"].isin(sites), [\"chain\", \"pdbSite\"]]\n",
    "    sel = \" or \".join([f\"c. {c} and i. {i}\" for _, c, i in sites.itertuples()])\n",
    "    col = DOMAIN_COLOR[domain]\n",
    "    pmlScript.extend([\n",
    "        f\"select {domain}, {sel}\",\n",
    "        f\"color {col}, {domain}\",\n",
    "        f\"set surface_color, {col}, {domain}\",\n",
    "        f\"set transparency, 0.75, {domain}\"\n",
    "#         f\"create Iso_{domain}, {domain}\",\n",
    "#         f\"show surface, Iso_{domain}\",\n",
    "#         f\"set surface_color, {col}, Iso_{domain}\",\n",
    "#         f\"set transparency, 0.1, Iso_{domain}\"\n",
    "    ])\n",
    "    \n",
    "# Script for parallel sites\n",
    "for category, sites in targetSites.items():\n",
    "    if category == \"parallel\":\n",
    "        sel = \" or \".join([f\"c. {c} and i. {i}\" for _, c, i in sites.itertuples()])\n",
    "        col = SITES_COLOR[category]\n",
    "        pmlScript.extend([\n",
    "            f\"select {category}, {sel}\",\n",
    "            f\"color {col}, {category}\",\n",
    "            f\"show sphere, {category}\",\n",
    "            f\"disable {category}\"\n",
    "        ])\n",
    "        pmlScript.append(\"reset\")\n",
    "        pmlScript.append(pmlSetView)\n",
    "        # For the site's label\n",
    "        label = \" or \".join([f\"c. {c} and i. {i} and n. CA\" for _, c, i in sites.itertuples()])\n",
    "        pmlScript.extend([\n",
    "            \"set label_position,(5,7,100)\",\n",
    "#             \"set label_size, 10\",\n",
    "            \"set label_connector, true\",\n",
    "            \"set label_bg_color, white\",\n",
    "            \"set label_bg_transparency, 0.2\",\n",
    "            f\"label {label}, resi\\n\"\n",
    "        ])\n",
    "\n",
    "with open(os.path.join(outdirPath, \"script_parallel.pml\"), 'w') as f:\n",
    "    f.write(\"; \".join(pmlScript))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
